### PR TITLE
Stake CW20 Rewards

### PR DESCRIPTION
**Pull request by Callum-A**
Picked up the TODOs left on PR #164.

- Added a query pending rewards route.
- Pulled in Main as the current stake-cw20-rewards was using an older version of the stake-cw20 contract. (That's why other commits are listed under this PR)
- Added integration tests
  - Checks initial config setup
  - Checks funding
  - Checks rewards
  - Checks compounding each block
  - Checks invalid claim rewards calls
 
Feedback appreciated.

### CODE REVIEW

**Comment by JakeHartnell** [comment]
Upstream PR needs a rebase, then it will be easier to review this. CC @ben2x4. :pray:

File: contracts/stake-cw20-rewards/src/contract.rs [hunk 1]
```diff
@@ -161,5 +163,486 @@ pub fn query_config(deps: Deps) -> StdResult<GetConfigResponse> {
     })
 }
 
+pub fn query_rewards(deps: Deps, env: Env) -> StdResult<GetPendingRewardsResponse> {
```
Comment (ezekiiel) [inline contracts/stake-cw20-rewards/src/contract.rs:166]: What do you think about renaming this to `query_pending_rewards`?

File: contracts/stake-cw20-rewards/src/contract.rs [hunk 2]
```diff
@@ -161,5 +163,486 @@ pub fn query_config(deps: Deps) -> StdResult<GetConfigResponse> {
...
+    let config = CONFIG.load(deps.storage)?;
+    let last_claim = LAST_CLAIM.load(deps.storage).unwrap_or(config.start_block);
+    let blocks = Uint128::from(min(&env.block.height, &config.end_block) - last_claim);
+    let pending_rewards = if config.funded {
+        blocks
+            .checked_mul(config.payment_per_block)
+            .map_err(StdError::overflow)?
+    } else {
+        Uint128::zero()
+    };
```
Comment (ezekiiel) [inline contracts/stake-cw20-rewards/src/contract.rs:176]: Does this contract have a `payment_block_delta` like the other one? If so I don't think this is quite accurate. If not, we should discuss if they should be consistent. https://github.com/DA0-DA0/dao-contracts/pull/170#discussion_r802210734

**Review by ezekiiel — COMMENTED** [review]
A rebase on main would be amazing for reviewing this 🙏

File: contracts/stake-cw20-rewards/src/contract.rs (hunk 2 above)
Comment (Callum-A) [reply contracts/stake-cw20-rewards/src/contract.rs:176]: This contract doesn't have a delta like external, as rewards are paid every block so I believe the maths here is correct in line with @ben2x4 's description in PR #164 but correct me if I'm misunderstanding.

File: contracts/stake-cw20-rewards/src/contract.rs (hunk 1 above)
Comment (Callum-A) [reply contracts/stake-cw20-rewards/src/contract.rs:166]: I agree this needs to be renamed.

**Comment by JakeHartnell** [comment]
@ezekiiel / @Callum-A, is this PR still relevant? It's been open for a while.

**Comment by JakeHartnell** [comment]
Yolo.

**Pull request merged**